src/omotes_orchestrator/timeout_job_manager.py

```diff
@@ -0,0 +1,109 @@
+
+LOGGER = logging.getLogger("omotes_orchestrator")
+
+
+class TimeoutJobManager:
+    """Periodically checks if any job is timed out. Cancel and delete the timed-out job."""
+
+    postgresql_if: PostgresInterface
+    """Interface to PostgreSQL."""
+    # orchestrator: Orchestrator  # TODO
+    """Orchestrator application instance"""
+    _init_time: datetime
+    """Instantiated datetime (UTC), is typically instantiated when the orchestrator is started"""
+    _active_threshold_sec: int
+    """Only when the instance is initialized longer than the threshold period
+    can it start canceling the timed-out jobs."""
+    _rerun_sec: int
+    """Period in seconds to rerun the timeout job cancellation task"""
+    _stop_event: threading.Event
+    """Event to signal the thread to stop"""
+    _timeout_jobs_handler_thread: threading.Thread
+    """Thread for canceling the timed-out job"""
+    _timeout_jobs_handler_thread_active: bool
+    """Flag indicating if the timeout job handler thread is active"""
+
+    def __init__(self, postgresql_if: PostgresInterface) -> None:
+        """Construct the timeout job manager."""
+        self.postgresql_if = postgresql_if
+        # self.orchestrator = orchestrator
+        self._init_time = datetime.now(timezone.utc)
+        # TODO: get this from config
+        self._active_threshold_sec = 20
+        # TODO: overwrite the value when finish
+        self._rerun_sec = 10
+
+        self._stop_event = threading.Event()
+        self._timeout_jobs_handler_thread = threading.Thread(target=self.timeout_jobs_handler)
+        self._timeout_jobs_handler_thread_active = False
+
+    def start(self) -> None:
+        """
+        TODO
+        """
+        if not self._timeout_jobs_handler_thread_active:
+            LOGGER.info("Starting the timeout job manager")
+
+            self._stop_event.clear()
+            self._timeout_jobs_handler_thread.start()
+            self._timeout_jobs_handler_thread_active = True
+
+    def stop(self) -> None:
+        """
+        TODO
+        """
+        if self._timeout_jobs_handler_thread_active:
+            LOGGER.info("Stopping the timeout job manager")
+
+            self._stop_event.set()
+            self._timeout_jobs_handler_thread.join(timeout=5.0)
+            self._timeout_jobs_handler_thread_active = False
+
+    def timeout_jobs_handler(self) -> None:
+        """
+        TODO
+        """
+        while not self._stop_event.is_set():
+            cur_time = datetime.now(timezone.utc)
+            active_sec = (cur_time - self._init_time).total_seconds()
+
+            if active_sec > self._active_threshold_sec:
+                jobs = self.postgresql_if.get_all_jobs()
+                for job in jobs:
+                    if self.job_is_timedout(job):
+                        # job_cancellation = JobCancel(uuid=str(job.job_id))
+                        
+                        # self.orchestrator.job_cancellation_handler(job_cancellation)
```

Comment:
@lfse-slafleur I made an initial implementation on issue #39
It now has a very similar setup to the `PostgresJobManager` component.

I want to reuse `job_cancellation_handler` from the orchestrator here. Also as it is mentioned in issue #39. The component should take `Orchestrator` as constructor arguments. However, this creates a circular import error. Any suggestions on this?

_This component should accept the PostgresInterface and Orchestrator as constructor arguments. The component MUST be isolated from other components in the orchestrator._

Reply:
- [5704e09](https://github.com/Project-OMOTES/orchestrator/pull/50/commits/5704e09abf744566450fd4079e5044ea5e01ba5f): Resolved circular import error based on the provided reference: https://stackoverflow.com/questions/39740632/python-type-hinting-without-cyclic-imports
- In the current implementation, `TimeoutJobManager` still takes `Orchestrator` as constructor arguments, but it is initialized with `None` in `main.py:main`. When `Orchestrator` is instantiated in `main.py:main`, the instance of Orchestrator will be assigned to `self.timeout_job_manager.orchestrator` to allow reusing of `job_cancellation_handler` method possible inside the TimeoutJobManager component.

   ```
        if self.timeout_job_manager.orchestrator is None:
            self.timeout_job_manager.orchestrator = self
            logger.info("Assigned orchestrator %s to the timeout job manager.", self)
   ```